crates/bevy_input/src/gamepad/types.rs

```diff
@@ -0,0 +1,147 @@
+];
+
+/// An array of every [`GamepadAxisType`].
+///
+/// ## Usage
+///
+/// This is used inside of the [`gamepad_event_system`](crate::gamepad::gamepad_event_system)
+/// to set or remove the values of every [`GamepadAxis`](crate::gamepad::GamepadAxis).
+pub(crate) const ALL_AXIS_TYPES: [GamepadAxisType; 8] = [
+    GamepadAxisType::LeftStickX,
+    GamepadAxisType::LeftStickY,
+    GamepadAxisType::LeftZ,
+    GamepadAxisType::RightStickX,
+    GamepadAxisType::RightStickY,
+    GamepadAxisType::RightZ,
+    GamepadAxisType::DPadX,
+    GamepadAxisType::DPadY,
+];
+
+/// A type of a [`GamepadButton`](crate::gamepad::GamepadButton).
+///
+/// ## Usage
+///
+/// This is used to determine which button has changed its value when receiving a
+/// [`GamepadEventType::ButtonChanged`].
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+pub enum GamepadButtonType {
+    /// The bottom action button of the action pad (i.e. PS: Cross, Xbox: A).
+    South,
+    /// The right action button of the action pad (i.e. PS: Circle, Xbox: B).
+    East,
+    /// The upper action button of the action pad (i.e. PS: Triangle, Xbox: Y).
+    North,
+    /// The left action button of the action pad (i.e. PS: Square, Xbox: X).
+    West,
+
+    /// The C button.
+    C,
+    /// The Z button.
+    Z,
+
+    /// The first left trigger.
+    LeftTrigger,
+    /// The second left trigger.
+    LeftTrigger2,
+    /// The first right trigger.
+    RightTrigger,
+    /// The second right trigger.
+    RightTrigger2,
+
+    /// The select button.
+    Select,
+    /// The start button.
+    Start,
+    /// The mode button.
+    Mode,
+
+    /// The left thumb stick button.
+    LeftThumb,
+    /// The right thumb stick button.
+    RightThumb,
+
+    /// The up button of the D-Pad.
+    DPadUp,
+    /// The down button of the D-Pad.
+    DPadDown,
+    /// The left button of the D-Pad.
+    DPadLeft,
+    /// The right button of the D-Pad.
+    DPadRight,
+}
+
+/// An type of a [`GamepadAxis`](crate::gamepad::GamepadAxis).
+///
+/// ## Usage
+///
+/// This is used to determine which axis has changed its value when receiving a
+/// [`GamepadEventType::ButtonChanged`].
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+pub enum GamepadAxisType {
+    /// The horizontal value of the left stick.
+    LeftStickX,
+    /// The vertical value of the left stick.
+    LeftStickY,
+    /// The value of the left `Z` button.
+    LeftZ,
+
+    /// The horizontal value of the right stick.
+    RightStickX,
+    /// The vertical value of the right stick.
+    RightStickY,
+    /// The value of the right `Z` button.
+    RightZ,
+
+    /// The horizontal value of the D-Pad.
+    DPadX,
+    /// The vertical value of the D-Pad.
+    DPadY,
+}
+
+/// A type of a [`GamepadEvent`](crate::gamepad::GamepadEvent).
```

Comment:
IMO we should probably just merge these two types into a single `enum`.

Reply:
Which two types exactly do you want to merge?
`GamepadButtonType` and `GamepadAxisType`?
If so do you have any name suggestions for the merged version?

Reply:
I meant we should merge `GamepadEvent` and `GamepadEventType`. That's a bit of a pain though, looking more closely, since we'd need to carry "which gamepad was this" everywhere :/ We can come back to this later.

Reply:
Interesting. Didn't think about this connection at all. Might be something to consider for the future. I'm also not 100% sold on the current implementation of our input handling.